orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishKeelSpecsTask.kt

```diff
@@ -0,0 +1,131 @@
+//        defaultValue: .netflix/
+//        order: 1
+//      - name: fileExtension
+//        mapping: cluster.env.MD_FILE_EXTENSION
+//        label: File Extension
+//        defaultValue: .yml
+//        order: 2
+//      - name: forceRepublish
+//        mapping: cluster.env.MD_FORCE_REPUBLISH
+//        defaultValue: false
+//        label: Force Republish
+//        order: 3
+//      - name: gitBranch
+//        mapping: cluster.env.GIT_BRANCH
+//        defaultValue: ${trigger.branch}
+//        label: Branch
+//        order: 4
+//      - name: gitSha
+//        mapping: cluster.env.GIT_SHA
+//        defaultValue: ${trigger.hash}
+//        label: SHA
+//        order: 5
+
+    val context = keelObjectMapper.convertValue<PublishKeelSpecsContext>(stage.context)
+    val manifests = scmService.listKeelManifests(
+      context.scmType, context.projectKey, context.repositorySlug, context.ref)
+
+    var failedManifest: String? = null
+    for (manifest in manifests) {
+      // TODO: do we even support multiple SCM servers of each kind? judging from igor, I'm guessing no
+      val uri = "${context.scmType}://<server???>/${context.projectKey}/${context.repositorySlug}/.netflix/spinnaker/$manifest"
+
+      log.info("Retrieving keel manifest at $uri")
+      // FIXME: I *think* the ref parameter below should actually be based on info from the (Rocket?) code event trigger
+      val manifestContents = scmService.getKeelManifest(
+        context.scmType, context.projectKey, context.repositorySlug, manifest, context.ref)
+
+      var resourceSpec: ResourceSpec
+      try {
+        resourceSpec = keelObjectMapper.convertValue(manifestContents)
+      } catch (e: java.lang.Exception) {
+        throw IllegalArgumentException("Unable to parse resource spec at $uri")
+      }
+
+      // TODO: confirm this is the right thing to do
+      val user = stage.execution.trigger.user ?: "anonymous"
+      log.info("Publishing manifest $manifest to keel on behalf of $user")
+      val response = keelService.publishSpec(resourceSpec, user)
+
+      if (!response.successful()) {
+        log.error("Failed to publish resource spec at $uri")
+        failedManifest = manifest
+        break
+      }
+    }
+
+    val outputs = mutableMapOf<String, Any>()
+    // TODO: add failed manifest info to output, if any
+    // TODO: add other outputs
+
+    val executionStatus = if (failedManifest == null) ExecutionStatus.SUCCEEDED else ExecutionStatus.TERMINAL
+    return TaskResult.builder(executionStatus).context(outputs).build()
+  }
+
+  override fun getBackoffPeriod() = TimeUnit.SECONDS.toMillis(15)
+
+  override fun getTimeout() = TimeUnit.MINUTES.toMillis(1)
+
+  private fun Response.successful() =
+    this.status in 200..299
+
+  data class PublishKeelSpecsContext(
+    val scmType: String,
+    val projectKey: String,
+    val repositorySlug: String,
+    val ref: String? = "refs/heads/master"
+    // TODO: forceRepublish -- what does it do?
```

Comment:
For the current stage we only publish manifests if the last commit contains changes to the directory they're in. If force republish is true, we always re-publish manifests.

Reply:
Got it! Thanks for explaining.